src/lib.rs

```diff
@@ -126,15 +126,49 @@ pub trait Renderable: Send + Sync {
     fn render(&self, context: &mut Context) -> Result<Option<String>>;
 }
 
+pub type TemplateLocation = Path;
+pub trait FileRepository {
+    fn read_template_file(&self, path: &TemplateLocation) -> Result<String>;
```

Comment:
I'm not sure what caching you're referring to in this case. 

If we're trying to avoid disk access, one could compose a `TemplateFactory` that has one FS layer and one in-memory layer for caching. 

If you're trying to avoid copying memory, we could be lending refs